spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,366 @@
+   val problem: TransformerType = LanguageModel)
+  (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+
+  override def buildModel(): Module[T] = {
+    problem match {
+      case LanguageModel => buildLM()
+      case Translation => buildTranslation()
+    }
+  }
+
+  private def buildTranslation(): Module[T] = {
+    // input: int tensor with shape [batch_size, input_length].
+    val inputNode = Input()
+    // target: int tensor with shape [batch_size, target_length].
+    val targetNode = Input()
+    val attentionBias = new PaddingMask().inputs(inputNode)
+
+    val join = JoinTable(1, -1).inputs(inputNode, targetNode)
+    val constantValue = math.sqrt(hiddenSize)
+    val embedding = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(join))
+    val split = new SplitTensor(1, 2).inputs(embedding)
+    val embeddingInput = SelectTable(1).inputs(split)
+    val embeddingOutput = SelectTable(2).inputs(split)
+
+    val encoderOutput = encode(embeddingInput, attentionBias)
+    val outputNode = decode(embeddingOutput, encoderOutput, attentionBias)
+    Graph(Array(inputNode, targetNode), outputNode)
+  }
+
+  private def buildLM(): Module[T] = {
+    val inputNode = Input()
+    val constantValue = math.sqrt(hiddenSize)
+    val embeddingInput = MulConstant(constantValue).inputs(
+      LookupTable[T](vocabSize, hiddenSize).inputs(inputNode))
+    val outputNode = decode(embeddingInput)
+    Graph(inputNode, outputNode)
+  }
+
+  private[nn] def encode(inputs: ModuleNode[T], attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    // Prepare inputs to the layer stack by adding positional encodings and
+    // applying dropout.
+    val position = new PositionEncode().inputs(inputs)
+    val encoderInput = CAddTable().inputs(inputs, position)
+    val decoderInputDrop = if (train) {
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(encoderInput)
+    } else encoderInput
+
+    encodeStack(numHiddenlayers, decoderInputDrop, attentionBias)
+  }
+
+  private[nn] def decode(targets: ModuleNode[T],
+                     encoderOutput: ModuleNode[T] = null,
+                     attentionBias: ModuleNode[T] = null): ModuleNode[T] = {
+    val decoderInput = new PositionEncodeWithShift().inputs(targets)
+    val decoderSelfAttentionBias = new SelfAttentionMask().inputs(targets)
+
+    val decoderInputDrop = if (train) {
+      val postDropOut = Dropout(1- postprocessDropout)
+      postDropOut.inputs(decoderInput)
+    } else decoderInput
+
+    decodeStack(numHiddenlayers, decoderInputDrop,
+      decoderSelfAttentionBias, encoderOutput, attentionBias)
+  }
+
+
+  private[nn] def encodeStack(numLayers: Int,
+                              encoderInput: ModuleNode[T],
+                              attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(numLayers, encoderInput, attentionBias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(numLayers: Int,
```

Comment:
seems this method is shared by both encoding and decoding, please rename to a common one, and make the `preName` as 'xxxType'

Reply:
ok, change to block, and make `preName` to 'blockType', which can be "encode" or "decode"